pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/builder/AbstractPartitionAwareRoutingTableBuilder.java

```diff
@@ -0,0 +1,120 @@
+ *      a. Check if the segment can be pruned. If pruned, go to the next segment.
+ *      b. If not pruned, assign the segment to a server with the replica id that is picked above.
+ *
+ */
+public abstract class AbstractPartitionAwareRoutingTableBuilder extends AbstractRoutingTableBuilder {
+  private static final Logger LOGGER = LoggerFactory.getLogger(AbstractPartitionAwareRoutingTableBuilder.class);
+
+  protected static final String PARTITION_METADATA_PRUNER = "PartitionZKMetadataPruner";
+  protected static final int NO_PARTITION_NUMBER = -1;
+
+  Map<SegmentId, Map<Integer, ServerInstance>> _segmentNameToServersMapping = new HashMap<>();
+  AtomicReference<Map<SegmentId, Map<Integer, ServerInstance>>> _mappingReference =
+      new AtomicReference<>(_segmentNameToServersMapping);
+  Map<SegmentId, SegmentZKMetadata> _segmentToZkMetadataMapping = new ConcurrentHashMap<>();
+  protected ZkHelixPropertyStore<ZNRecord> _propertyStore;
+  protected SegmentZKMetadataPrunerService _pruner;
+  protected TableConfig _tableConfig;
+  protected Random _random = new Random();
+  protected int _numReplicas;
+
+  @Override
+  public void init(Configuration configuration, TableConfig tableConfig, ZkHelixPropertyStore<ZNRecord> propertyStore) {
+    _tableConfig = tableConfig;
+    _propertyStore = propertyStore;
+    _pruner = new SegmentZKMetadataPrunerService(new String[]{PARTITION_METADATA_PRUNER});
+  }
+
+  @Override
+  public Map<ServerInstance, SegmentIdSet> findServers(RoutingTableLookupRequest request) {
+    Map<ServerInstance, SegmentIdSet> result = new HashMap<>();
+    Map<SegmentId, Map<Integer, ServerInstance>> mappingReference = _mappingReference.get();
+    SegmentPrunerContext prunerContext = new SegmentPrunerContext(request.getBrokerRequest());
+
+    // 1. Randomly pick a replica id
+    int replicaGroupId = _random.nextInt(_numReplicas);
+    for (SegmentId segmentId : mappingReference.keySet()) {
+      SegmentZKMetadata segmentZKMetadata = _segmentToZkMetadataMapping.get(segmentId);
+
+      // 2a. Check if the segment can be pruned
+      boolean segmentPruned = (segmentZKMetadata != null) && _pruner.prune(segmentZKMetadata, prunerContext);
+
+      if (!segmentPruned) {
+        // 2b. Segment cannot be pruned. Assign the segment to a server with the replica id picked above.
+        Map<Integer, ServerInstance> replicaId2ServerMapping = mappingReference.get(segmentId);
+        ServerInstance serverInstance = replicaId2ServerMapping.get(replicaGroupId);
+
+        // When the server is not available with this replica id, we need to pick another available server.
+        if (serverInstance == null) {
+          if (!replicaId2ServerMapping.isEmpty()) {
+            serverInstance = replicaId2ServerMapping.values().iterator().next();
+          } else {
+            // No server is found for this segment.
+            LOGGER.warn("No server is found for the segment {}.", segmentId.getSegmentId());
```

Comment:
Should we log a metric for this? Also, should this be sent back as partial response?